### ISSUE TITLE
Steps towards plugins repository

### ISSUE THREAD

**Issue by Technius** [xournalpp/xournalpp#3839]
Currently, some "community plugins" are packaged with the application, but they should be split off into a separate `xournalpp-plugins` repository as they are orthogonal to the core application functionality. This issue tracks the steps needed before we can do this split off.

### Pros and cons of separate plugins repository

Pros:

1. It provides a focused place for people to discuss/contribute plugins and makes the review process less troublesome.
2. I do not like bundling so many plugins with the application, which defeats the purpose of making them "plugins"
3. Including the plugins with the application means that we have to maintain the plugins as part of each new version of the application
4. The separation allows us to complicating the copyright of the main project, as plugins may use 3rd party code/resources.
5. Users can still install the community plugins manually if they want them (the Windows installer could even bundle them and offer to install the plugins with a checkbox).

Cons:

1. There is currently no sort of "supported Xournal++ versions" or anything like that. Unclear what "plugin API compatibility" is available. Suggested mitigation: readme on xournalpp-plugins that lists compatibility.
2. Plugin API is still very incomplete and unstable, and the most useful features are only available on nightly builds.
3. Users that rely on the community plugins as part of their daily workflow will no longer have them available by default
4. Probably more cons...

### Tasks

* [ ] Create xournalpp-plugins repository
* [ ] Implement custom plugin folder (#2422 or #3063 or a new PR)
* [ ] Move community plugins to xournalpp-plugins

**Comment by ShadowDara** [comment]
The supported versions if of course a big issue for a pluigin hub / install manager

If think i have an solution for this, but this is at the moment only the repository hub.

---

[https://github.com/ShadowDara/xournalpp-pluigin-hub-idea](https://github.com/ShadowDara/xournalpp-pluigin-hub-idea)

This Repository is a template for the Hub and a i added a template/exampl pluigin

---

### The Plan:

1. a user registrates a new pluigin with a json file in `pluigin-registration`
2. an automatic workflow will collect the data about the pluigin
3. the workflow will create a json file in `pluigin` with the same namespace

the json file in `pluigin` could look sth like this:

```json
{
    "name": "",
    "author": "",
    "description": "",
    "versions": [
        {
            "available_for": [],
            "link": ""
        }
    ]
}
```

---

### Some Problems could be
- a very high workflow load which would be very expensive
- another solution would that the users are self creating the json file, the problem here would be to register a new version, users have always to make a PR

---

If you like this idea, i would continue develeping this and then trying to implement a downloader directly to Xournall++

Through a pluigin, this would be really easy, but it is a good thing that pluigin have no internet access for security reasons

---

~~Not really sure at the moment if this is possible but maybe add internet access for pluigins for hardcoded adresses?~~

**Comment by ShadowDara** [comment]
> Not really sure at the moment if this is possible but maybe add internet access for pluigins for hardcoded adresses?

No that doesn't work, because you also need a pluigin downloader and these github repositories are distributed on their user accounts

**Comment by ShadowDara** [comment]
Hey, I hope I’m not bothering you. Just wondering if you've had a moment to look at my idea or if you think it still needs improvement?

**Comment by bhennion** [comment]
Thanks for your proposal. Overall, a plugin repository is mainly independent of Xournal++ itself. Seeing how streched thin the Xournal++ team already is, I think this should be developed independently. In that case, the structure is really up to you (or anyone else who implements it).

Ideally, I think the implementation should also provide a client (i.e. a plugin manager). It could easily be a separate app (that we would ship with Xournal++) - so that you are completely free in the protocols and all that.

**Comment by ShadowDara** [comment]
Thank your for your answer,

this sounds good, so i will continue the project

for anyone who is reading and want to help, look at my current work, here:
[ShadowDara/xournalpp-pluigin-hub-idea](https://github.com/ShadowDara/xournalpp-pluigin-hub-idea)

**Comment by ShadowDara** [comment]
Hey, its me again. Its been quite a while to be honest. I've now finished my plan for a good package manager which would although fit Xournal++ and now for the Implementation, i just want to ask. For simplicity reasons i am tending to not write it in C++. Would this be allowed? Personally i am more tending to Go or Rust.

**Comment by rolandlo** [comment]
@ShadowDara Can you share more details on how your plugin hub is going to work. Am I correct that:
- There is a Github repository for plugin authors to share their plugin
- Plugin authors will make PR's for getting their plugin into that repository or updating the supported version numbers
- There is a standalone plugin viewer app and a website where you can browse through plugins (from that repository) and download those you want
- The plugin manager in Xournal++ would continue to be used solely for activating and deactivating plugins

For a standalone project I think using a different language than C++ would be acceptable.
